src/core/geometry/qgssphere.h

```diff
@@ -0,0 +1,188 @@
+/***************************************************************************
+                         qgssphere.h
+                         --------------
+    begin                : July 2023
+    copyright            : (C) 2023 by Nyall Dawson
+    email                : nyall dot dawson at gmail dot com
+ ***************************************************************************/
+
+/***************************************************************************
+ *                                                                         *
+ *   This program is free software; you can redistribute it and/or modify  *
+ *   it under the terms of the GNU General Public License as published by  *
+ *   the Free Software Foundation; either version 2 of the License, or     *
+ *   (at your option) any later version.                                   *
+ *                                                                         *
+ ***************************************************************************/
+
+#ifndef QGSSPHERE_H
+#define QGSSPHERE_H
+
+#include "qgis_core.h"
+#include "qgis_sip.h"
+#include "qgis.h"
+#include <limits>
+
+class QgsPoint;
+class QgsCircle;
+class QgsBox3d;
+
+/**
+ * \ingroup core
+ * \class QgsSphere
+ * \brief A spherical geometry object.
+ *
+ * Represents a simple 3-dimensional sphere.
+ *
+ * \since QGIS 3.34
+ */
+class CORE_EXPORT QgsSphere
+{
+
+  public:
+
+    /**
+     * Constructor for an invalid QgsSphere.
+     */
+    QgsSphere() SIP_HOLDGIL = default;
+
+    /**
+     * Constructor for QgsSphere with the specified center (\a x, \a y, \a z) and \a radius.
+     */
+    QgsSphere( double x, double y, double z, double radius ) SIP_HOLDGIL;
+
+    bool operator==( const QgsSphere &other ) const
+    {
+      return qgsDoubleNear( mCenterX, other.mCenterX ) && qgsDoubleNear( mCenterY, other.mCenterY ) && qgsDoubleNear( mCenterZ, other.mCenterZ ) && qgsDoubleNear( mRadius, other.mRadius );
+    }
+    bool operator!=( const QgsSphere &other ) const { return !( *this == other ); }
+
+    /**
+     * Returns TRUE if the sphere is a null (default constructed) sphere.
+     */
+    bool isNull() const SIP_HOLDGIL;
+
+    /**
+     * Returns TRUE if the sphere is considered empty, i.e. it has a radius of 0.
+     */
+    bool isEmpty() const SIP_HOLDGIL;
+
+    /**
+     * Returns the center point of the sphere.
+     *
+     * \see centerX()
+     * \see centerY()
+     * \see centerZ()
+     * \see setCenter()
+     */
+    QgsPoint center() const SIP_HOLDGIL;
```

Comment:
Could we use QgsVector3D instead for center()/setCenter()? It is a leaner object and imho better for some math operations (no worrying about M, no virtual methods...)

Reply:
I can, yes. (I'm not using QgsPoint in the class itself though for this same reason, these are just convenience functions vs the seperate double getters)

Reply:
I've added `QgsVector3D centerVector() const` for this use case now.